### PR TITLE
Reworked the Results model to to generalize a few functions and have graceful errors

### PR DESCRIPTION
**Pull request by Lnaden**
As it says on the tin

Added in an Optimization Proceduere for our first procedure action.

There will need to be work to be more general later, but this should be good for now

### CODE REVIEW

File: qcelemental/models/common_models.py [hunk 1]
```diff
@@ -8,3 +9,34 @@ class Provenance(BaseModel):
...
+        allow_extra = True
+
+
+class DriverEnum(str, Enum):
+    energy = 'energy'
+    gradient = 'gradient'
+    hessian = 'hessian'
+
+
+class QCEngineError(BaseModel):
```
Comment (dgasmith) [inline qcelemental/models/common_models.py]: Should just be a `ComputeError` not specific to QCE.

File: qcelemental/models/procedures.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                        regex=qcschema_optimization_output_default) = qcschema_optimization_output_default
+    final_molecule: Molecule
+    trajectory: List[Result] = None
+    energies: Union[List[float], float] = None
+    success: bool
+    error: QCEngineError = None
+
+
+
+    __json_mapper = {
```
Comment (dgasmith) [inline qcelemental/models/procedures.py]: Can we remove everything below this line, should not be here.

File: qcelemental/models/procedures.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    success: bool = False
+    error: QCEngineError
+
+
+class Optimization(FailedOptimization):
+    schema_name: constr(strip_whitespace=True,
+                        regex=qcschema_optimization_output_default) = qcschema_optimization_output_default
+    final_molecule: Molecule
+    trajectory: List[Result] = None
+    energies: Union[List[float], float] = None
```
Comment (dgasmith) [inline qcelemental/models/procedures.py]: `energies` will always be a `List[float]`.

File: qcelemental/models/procedures.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
+from pydantic import BaseModel, constr
+from .common_models import Model, DriverEnum, \
```
Comment (dgasmith) [inline qcelemental/models/procedures.py]: Canonical way to do this is with parenthesis `from .common_models import (Model, DriverEnum ...)`.

File: qcelemental/models/results.py [hunk 5]
```diff
@@ -68,21 +47,64 @@ class ResultInput(BaseModel):
...
     schema_version: int = 1
     keywords: dict = {}
     provenance: Provenance = provenance_stamp(__name__)
 
     class Config:
         allow_mutation = False
+        allow_extra = True
+
+
+class FailedResult(ResultInput):
```
Comment (dgasmith) [inline qcelemental/models/results.py]: I think these Failed classes are a bit much. Best to have a single `ComputeError` class where we have:
```python
success = False
input_data = Any
error_type = str
error_message = str
```

**Comment by dgasmith** [comment]
Thinking through that a bit more in general we will only have that particular return error come up if there is something wrong with the overall JSON structure. By default we will be returning a `Optimization` or `Result` with `success = False` 99.9% of the time with a compute error in its current form.

Perhaps we should leave the ComputeError as is and build out a `UnrecognizedModel` field or similar? Where the creation of the object went wrong.

**Comment by Lnaden** [comment]
> Perhaps we should leave the ComputeError as is and build out a UnrecognizedModel field or similar? Where the creation of the object went wrong.

This is part of the reason why I made the objects in the first place, they keep a pydantic model, but can accept everything, this way, we can treat all the post-compute processing (e.g. the `handle_output_metadata`) the same way since the same object structure will always be returned (Pydantic objects) without being lossy

**Comment by Lnaden** [comment]
I also made the `FailedResult` and `FailedOptimization` objects so that they would all at least have the correct attributes, even if they wound up being `None`

**Comment by dgasmith** [comment]
Then we get an object that is very dangerous, it quacks like a Result but isn't. Anything that fails that hard should be very clear it isn't a valid object.

**Comment by Lnaden** [comment]
> Then we get an object that is very dangerous, it quacks like a Result but isn't. Anything that fails that hard should be very clear it isn't a valid object.

Fair. Would you be okay with something like this:

```python
class ComputeError(BaseModel):
    error_type: str  
    error_message: str

     class Config:
        allow_extra = False

class FailedOperation(BaseModel):
    success: bool = False
    error: ComputeError

    class Config:
        allow_extra = True
        allow_mutable = False

class FailedResult(FailedOperation):
    pass

class FailedOptimization(FailedOperation):
    pass
```

This way they are all common objects, but we can infer where the failure happened by name? Or would you prefer just the `FailedOperation` object?

**Comment by dgasmith** [comment]
There isn't a reason anywhere in the current code where we would need a `FailedResult`. Can you describe a use case for it?

**Comment by Lnaden** [comment]
At the moment, I'm only trapping malformed inputs from `dict` type input, and I skipped over the `except Exception as e:` since we agreed on speed now, refine later. The use case would be if the `get_program(program)(input_data, config)` call in QCEngine failed for whatever reason: program crashes, user provides really dumb inputs which are technically "valid," this would be the catchall.

**Comment by Lnaden** [comment]
The other use case right now is program does not exist, which is possible. I'm fine only having a `FailedOperation` class, but we need some way to pass back a Pydantic object which does no validation except on the error, otherwise we are going to have a ton of additional `try: except` clauses in the code as we try to read back in data and many `if isinstance(obj, dict)` since internally we assume we're going to have the Pydantic models, unless there is an error, then we would need a different set of functional calls since we cant do `{BaseModel}['some_attribute']` and have it work

**Comment by dgasmith** [comment]
Right, which is why I favor `FailedOperation`. But why do we need `FailedResult/FailedOptimization`?

**Comment by Lnaden** [comment]
>  But why do we need FailedResult/FailedOptimization?

We don't, per se, I just was considering making them so we could distingush between Results and Procedure objects since we have no safe/reliable way to separate them unless we inspect the internal attributes and guess

**Comment by Lnaden** [comment]
The counter argument would be results will only ever be expected from certain functions, and procedures from others, so the distinction would be automatic on its own.

**Comment by dgasmith** [comment]
The code in general does not handle failed results at that level. If it worked far enough to get a `Result/Optimization` object with `success = False` we may do something. Everything else is considered not worthwhile.

This result is just one very small piece of puzzle and QCFractal is very aware what each task is. Once the data is turned back into a dict it looses the class description anyway.

**Comment by Lnaden** [comment]
> If it worked far enough to get a Result/Optimization object with success = False we may do something. Everything else is considered not worthwhile.

Kind of, in some cases, it only returned the raw "success/error" dict. So having a "failed" class accept anything with those mandatory would work. 

>  Once the data is turned back into a dict it looses the class description anyway.

Fair. Are you still okay with a result, on any error, returning a `FailedOperation` with the input just fed in as additional attributes (although feeding into a `inputs: input_dict` would be safer). Or do you just want to return a `Failed` object which drops all of the input information?

**Comment by Lnaden** [comment]
I think I have all the requests you made in. Let me know what you think. If you want to make any more changes tonight, feel free to do so as I'll be stepping away for the night barring a disaster (call me if need be)

**Comment by dgasmith** [comment]
Well, definitely will not be calling anyone about code emergencies. We are not SysAdmins ;0

Changes look great, thanks for making them.

**Comment by Lnaden** [comment]
> We are not SysAdmins ;0

Now then, you *say* that....

> thanks for making them.
+:praisethesun:

**Pull request merged**